test/functional/transactions_tests.js

```diff
@@ -74,34 +76,81 @@ function translateClientOptions(options) {
   return options;
 }
 
-// Main test runner
-describe('Transactions (spec)', function() {
-  const testContext = {
-    dbName: 'transaction-tests',
-    collectionName: 'test'
-  };
-
-  const testSuites = fs
-    .readdirSync(`${__dirname}/spec/transactions`)
+function gatherTestSuites(specPath) {
+  return fs
+    .readdirSync(specPath)
     .filter(x => x.indexOf('.json') !== -1)
     .map(x =>
-      Object.assign(JSON.parse(fs.readFileSync(`${__dirname}/spec/transactions/${x}`)), {
+      Object.assign(JSON.parse(fs.readFileSync(`${specPath}/${x}`)), {
         name: path.basename(x, '.json')
       })
     );
+}
+
+class TransactionsTestContext {
+  constructor() {
+    this.url = null;
+    this.sharedClient = null;
+  }
+
+  setup(config) {
+    this.url = `mongodb://${config.host}:${config.port}/?replicaSet=${config.replicasetName}`;
+
+    this.sharedClient = config.newClient(this.url);
+    return this.sharedClient.connect();
+  }
+
+  teardown() {
+    return this.sharedClient.close();
+  }
+}
 
-  after(() => testContext.sharedClient.close());
-  before(function() {
-    // create a shared client for admin tasks
-    const config = this.configuration;
-    testContext.url = `mongodb://${config.host}:${config.port}/${testContext.dbName}?replicaSet=${
-      config.replicasetName
-    }`;
+// Main test runner
+describe('Transactions', function() {
+  const testContext = new TransactionsTestContext();
+
+  [
+    { name: 'spec tests', specPath: `${__dirname}/spec/transactions` },
+    {
+      name: 'withTransaction spec tests',
+      specPath: `${__dirname}/spec/transactions/convenient-api`
+    }
+  ].forEach(suiteSpec => {
+    describe(suiteSpec.name, function() {
+      const testSuites = gatherTestSuites(suiteSpec.specPath);
+      after(() => testContext.teardown());
+      before(function() {
+        return testContext.setup(this.configuration);
+      });
 
-    testContext.sharedClient = config.newClient(testContext.url);
-    return testContext.sharedClient.connect();
+      generateTestSuiteTests(testSuites, testContext);
+    });
   });
 
+  describe('withTransaction', function() {
+    let session;
+    beforeEach(() => {
+      const topology = new core.Server();
+      const sessionPool = new sessions.ServerSessionPool(topology);
+      session = new sessions.ClientSession(topology, sessionPool);
+    });
+
+    it('should provide a useful error if a Promise is not returned', {
+      metadata: { requires: { topology: ['replicaset', 'mongos'], mongodb: '>=4.0.x' } },
+      test: function() {
+        function fnThatDoesntReturnPromise() {
+          return false;
+        }
+
+        expect(() => session.withTransaction(fnThatDoesntReturnPromise)).to.throw(
```

Comment:
I may have missed this with the core review, but does this test mean that core is throwing a raw string, or does the chai helper just do magic to compare the message?

Reply:
The latter; chai compares the regex to the `message` property of the error subclass.